core/src/main/java/hudson/model/Run.java

```diff
@@ -1159,6 +1181,30 @@ private int addArtifacts(@Nonnull VirtualFile dir,
     public static final int TREE_CUTOFF = Integer.parseInt(SystemProperties.getString("hudson.model.Run.ArtifactList.treeCutoff", "40"));
 
     // ..and then "too many"
+    
+    /** {@link Run.Artifact} without the implicit link to {@link Run} */
+    private static final class SerializableArtifact implements Serializable {
+        private static final long serialVersionUID = 1L;
+        final String name;
+        final String relativePath;
+        final String href;
+        final String length;
+        final String treeNodeId;
+        SerializableArtifact(String name, String relativePath, String href, String length, String treeNodeId) {
+            this.name = name;
+            this.relativePath = relativePath;
+            this.href = href;
+            this.length = length;
+            this.treeNodeId = treeNodeId;
+        }
+    }
+
+    /** {@link Run.ArtifactList} without the implicit link to {@link Run} */
+    private static final class SerializableArtifactList extends ArrayList<SerializableArtifact> {
+        private static final long serialVersionUID = 1L;
+        private LinkedHashMap<SerializableArtifact, String> tree = new LinkedHashMap<>();
```

Comment:
Even for a private class not looks a bit dangerous since you do not override modification methods to put data to both containers (array and tree)

Reply:
This code was already a mess and I am trying to touch it as little as possible. The problem here was that `Artifact` and `ArtifactList` were, surprisingly, non`static`—they held a `Run` reference—which is unsupportable when they must be constructed inside a callable which _might_ run remotely. (If using the `FilePath`-based `VirtualFile`, for example. In fact this is not currently used by any `ArtifactManager`, only workspace browsing, but the API does not preclude the possibility, so…)